File: frontend/src/common/entities.ts
import { CONSORTIA } from "src/components/CreateCollectionModal/components/Content/common/constants";

export enum COLLECTION_LINK_TYPE {
  DOI = "DOI",
  RAW_DATA = "RAW_DATA",
  PROTOCOL = "PROTOCOL",
  LAB_WEBSITE = "LAB_WEBSITE",
  OTHER = "OTHER",
  DATA_SOURCE = "DATA_SOURCE",
}

/**
 * Options backing link type menu, used when associating links with collection.
 */
export const COLLECTION_LINK_TYPE_OPTIONS = {
  [COLLECTION_LINK_TYPE.DOI]: {
    text: "Publication DOI",
    value: COLLECTION_LINK_TYPE.DOI,
  },
  [COLLECTION_LINK_TYPE.RAW_DATA]: {
    text: "Raw Data",
    value: COLLECTION_LINK_TYPE.RAW_DATA,
  },
  [COLLECTION_LINK_TYPE.PROTOCOL]: {
    text: "Protocol",
    value: COLLECTION_LINK_TYPE.PROTOCOL,
  },
  [COLLECTION_LINK_TYPE.LAB_WEBSITE]: {
    text: "Lab Website",
    value: COLLECTION_LINK_TYPE.LAB_WEBSITE,
  },
  [COLLECTION_LINK_TYPE.OTHER]: {
    text: "Other",
    value: COLLECTION_LINK_TYPE.OTHER,
  },
  [COLLECTION_LINK_TYPE.DATA_SOURCE]: {
    text: "Data Source",
    value: COLLECTION_LINK_TYPE.DATA_SOURCE,
  },
};

/**
 * Collection status, used to determine whether a collection is private or public, or in revision.
 */
export enum COLLECTION_STATUS {
  PRIVATE = "Private",
  PUBLISHED = "Published",
  REVISION = "Revision",
}

export interface Link {
  link_name: string;
  link_url: string;
  link_type: COLLECTION_LINK_TYPE;
}

export enum ACCESS_TYPE {
  READ = "READ",
  WRITE = "WRITE",
}

export enum VISIBILITY_TYPE {
  PUBLIC = "PUBLIC",
  PRIVATE = "PRIVATE",
}

export enum X_APPROXIMATE_DISTRIBUTION {
  COUNT = "count",
  NORMAL = "normal",
}

export enum IS_PRIMARY_DATA {
  PRIMARY = "PRIMARY",
  SECONDARY = "SECONDARY",
  BOTH = "BOTH",
}

/**
 * Author of publication associated with a collection, populated from Crossref as part of collection publication
 * metadata.
 */
export interface Author {
  family: string;
  given: string;
}

export interface Collection {
  access_type: ACCESS_TYPE;
  consortia: CONSORTIA[];
  contact_email: string;
  contact_name: string;
  description: string;
  id: string;
  organs: string[];
  name: string;
  owner: string;
  visibility: VISIBILITY_TYPE;
  datasets: Map<Dataset["id"], Dataset>;
  links: Link[];
  data_submission_policy_version: string;
  created_at: number;
  updated_at: number;
  publisher_metadata: PublisherMetadata;
  revision_diff: boolean;
  summaryCitation?: string;
  tombstone?: boolean;
  revising_in?: Collection["id"];
  revision_of?: Collection["id"];
}

/**
 * Consortium of publication associated with a collection, populated from Crossref as part of collection publication
 * metadata.
 */
export interface Consortium {
  name: string;
}

export type Ontology = {
  label: string;
  ontology_term_id: string;
};

export interface Dataset {
  id: string;
  assay: Ontology[];
  tissue: Ontology[];
  disease: Ontology[];
  cell_count: number | null;
  // sex: string;
  self_reported_ethnicity: Ontology;
  organism: Ontology[];
  name: string;
  cell_type: Ontology[];
  is_primary_data: IS_PRIMARY_DATA;
  x_approximate_distribution?: X_APPROXIMATE_DISTRIBUTION;
  schema_version: string;
  // source_data_location: string;
  // revision: number;
  dataset_deployments: DatasetDeployment[];
  dataset_assets: DatasetAsset[];
  processing_status: DatasetUploadStatus;
  collection_id: Collection["id"];
  // contributors: Contributor[];
  // preprint_doi: DOI;
  // publication_doi: DOI;
  created_at: number;
  original_id?: string;
  published?: boolean;
  published_at?: number;
  tombstone?: boolean;
  updated?: boolean;
  collection_visibility: Collection["visibility"];
}

export enum DATASET_ASSET_FORMAT {
  H5AD = "H5AD",
  RDS = "RDS",
  CXG = "CXG",
}

export enum DATASET_ASSET_TYPE {
  REMIX = "REMIX",
  ORIGINAL = "ORIGINAL",
}

export interface DatasetAsset {
  id: string;
  dataset_id: string;
  filetype: DATASET_ASSET_FORMAT;
  filename: string;
  type: DATASET_ASSET_TYPE;
  s3_uri: string;
}

export interface DatasetDeployment {
  url: string;
}

// interface Contributor {
//   id: string;
//   name: string;
//   email: string;
//   institution: string;
// }

// interface DOI {
//   id: string;
//   title: string;
//   data: string;
//   journal: string;
// }

export interface User {
  picture: string;
}

export enum UPLOAD_STATUS {
  WAITING = "WAITING",
  UPLOADING = "UPLOADING",
  UPLOADED = "UPLOADED",
  FAILED = "FAILED",
  CANCEL_PENDING = "CANCEL_PENDING",
  PENDING = "PENDING",
  CANCELED = "CANCELED",
  NA = "NA",
}

export enum VALIDATION_STATUS {
  VALIDATING = "VALIDATING",
  VALID = "VALID",
  INVALID = "INVALID",
  NA = "NA",
}

export enum CONVERSION_STATUS {
  CONVERTING = "CONVERTING",
  CONVERTED = "CONVERTED",
  FAILED = "FAILED",
  SKIPPED = "SKIPPED",
  NA = "NA",
}

export enum PROCESSING_STATUS {
  PENDING = "PENDING",
  SUCCESS = "SUCCESS",
  FAILURE = "FAILURE",
}

export interface DatasetUploadStatus {
  dataset_id: string;
  upload_status: UPLOAD_STATUS;
  processing_status: PROCESSING_STATUS;
  upload_message: string;
  upload_progress: number;
  validation_status: VALIDATION_STATUS;
  validation_message: string;
  h5ad_status: CONVERSION_STATUS;
  cxg_status: CONVERSION_STATUS;
  rds_status: CONVERSION_STATUS;
}

export interface GeneSet {
  name: string;
  id: string;
  description: string;
  linked_datasets: Dataset["id"][];
}

/**
 * Collection publication metadata, populated from Crossref by collection publication DOI.
 */
export interface PublisherMetadata {
  authors: (Author | Consortium)[];
  journal: string;
  published_at: number;
  published_day: number;
  published_month: number;
  published_year: number;
}

File: backend/layers/common/entities.py
import uuid
from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from typing import List, Optional
from urllib.parse import urlparse

from dataclasses_json import dataclass_json

# TODO: copy and paste the docs for these


class DatasetStatusKey(str, Enum):
    UPLOAD = "upload"
    VALIDATION = "validation"
    CXG = "cxg"
    RDS = "rds"
    H5AD = "h5ad"
    PROCESSING = "processing"


class DatasetStatusGeneric:
    pass


class DatasetProcessingStatus(DatasetStatusGeneric, Enum):
    """
    Enumerates the status of processing a dataset.

    INITIALIZED = Dataset id created, and awaiting upload.
    PENDING = Processing has not started
    SUCCESS = Processing succeeded
    FAILURE = Processing failed
    """

    INITIALIZED = "INITIALIZED"
    PENDING = "PENDING"
    SUCCESS = "SUCCESS"
    FAILURE = "FAILURE"


class DatasetUploadStatus(DatasetStatusGeneric, Enum):
    NA = "NA"
    WAITING = "WAITING"
    UPLOADING = "UPLOADING"
    UPLOADED = "UPLOADED"
    FAILED = "FAILED"
    CANCEL_PENDING = "CANCEL PENDING"
    CANCELED = "CANCELED"


class DatasetValidationStatus(DatasetStatusGeneric, Enum):
    NA = "NA"
    VALIDATING = "VALIDATING"
    VALID = "VALID"
    INVALID = "INVALID"


class DatasetConversionStatus(DatasetStatusGeneric, Enum):
    NA = "NA"
    CONVERTING = "CONVERTING"
    CONVERTED = "CONVERTED"
    UPLOADING = "UPLOADING"
    UPLOADED = "UPLOADED"
    FAILED = "FAILED"
    SKIPPED = "SKIPPED"


class CollectionLinkType(str, Enum):
    DOI = "doi"
    RAW_DATA = "raw_data"
    PROTOCOL = "protocol"
    LAB_WEBSITE = "lab_website"
    OTHER = "other"
    DATA_SOURCE = "data_source"


class DatasetArtifactType(str, Enum):
    RAW_H5AD = "raw_h5ad"
    H5AD = "h5ad"
    RDS = "rds"
    CXG = "cxg"


class CollectionVisibility(Enum):
    """
    Describes a DbCollection's visibility.
    At most, one LIVE and one EDIT entry of a Collection may exist at a time.

    PUBLIC - a published and publicly viewable Collection.
    PRIVATE - an open Submission, i.e an unpublished and non-public Collection.
    """

    PUBLIC = "Public"
    PRIVATE = "Private"


@dataclass_json
@dataclass
class DatasetStatus:
    upload_status: Optional[DatasetUploadStatus]
    validation_status: Optional[DatasetValidationStatus]
    cxg_status: Optional[DatasetConversionStatus]
    rds_status: Optional[DatasetConversionStatus]
    h5ad_status: Optional[DatasetConversionStatus]
    processing_status: Optional[DatasetProcessingStatus]
    validation_message: Optional[str] = None

    @staticmethod
    def empty():
        return DatasetStatus(None, None, None, None, None, None)


@dataclass
class EntityId:
    id: str

    def __init__(self, entity_id: str = None):
        self.id = str(entity_id) if entity_id is not None else str(uuid.uuid4())

    def __repr__(self) -> str:
        return self.id


class CollectionId(EntityId):
    pass


class CollectionVersionId(EntityId):
    pass


class DatasetId(EntityId):
    pass


class DatasetVersionId(EntityId):
    pass


class DatasetArtifactId(EntityId):
    pass


@dataclass
class DatasetArtifact:
    id: DatasetArtifactId
    type: DatasetArtifactType
    uri: str

    def get_file_name(self):
        return urlparse(self.uri).path.split("/")[-1]


@dataclass
class OntologyTermId:
    label: str
    ontology_term_id: str


@dataclass_json
@dataclass
class DatasetMetadata:
    name: str
    schema_version: str
    organism: List[OntologyTermId]
    tissue: List[OntologyTermId]
    assay: List[OntologyTermId]
    disease: List[OntologyTermId]
    sex: List[OntologyTermId]
    self_reported_ethnicity: List[OntologyTermId]
    development_stage: List[OntologyTermId]
    cell_type: List[OntologyTermId]
    cell_count: int
    mean_genes_per_cell: float
    batch_condition: List[str]
    suspension_type: List[str]
    donor_id: List[str]
    is_primary_data: str
    x_approximate_distribution: Optional[str]
    primary_cell_count: Optional[int] = None
    feature_count: Optional[int]
    feature_biotype: Optional[List[str]]
    feature_reference: Optional[List[str]]
    default_embedding: Optional[str]
    embeddings: Optional[List[str]]


@dataclass
class CanonicalDataset:
    dataset_id: DatasetId
    dataset_version_id: Optional[DatasetVersionId]
    tombstoned: bool
    published_at: Optional[datetime] = None
    revised_at: Optional[datetime] = None  # The last time this Dataset Version was Published


@dataclass
class DatasetVersion:
    dataset_id: DatasetId
    version_id: DatasetVersionId
    collection_id: CollectionId  # Pointer to the canonical collection id this dataset belongs to
    status: DatasetStatus
    metadata: Optional[DatasetMetadata]
    artifacts: List[DatasetArtifact]
    created_at: datetime
    canonical_dataset: CanonicalDataset


@dataclass
class PublishedDatasetVersion(DatasetVersion):
    collection_version_id: CollectionVersionId  # Pointer to collection version it was originally published under
    published_at: datetime
    revised_at: datetime = None


@dataclass
class Link:
    name: Optional[str]
    type: str
    uri: str

    def strip_fields(self):
        if self.name:
            self.name = self.name.strip()
        self.type = self.type.strip()
        self.uri = self.uri.strip()


@dataclass_json
@dataclass
class CollectionMetadata:
    name: str
    description: str
    contact_name: str
    contact_email: str
    links: List[Link]
    consortia: List[str] = field(default_factory=list)


@dataclass
class CanonicalCollection:
    id: CollectionId
    version_id: Optional[CollectionVersionId]  # Needs to be optional, or not exist
    originally_published_at: Optional[datetime]
    revised_at: Optional[datetime]
    tombstoned: bool


@dataclass
class CollectionVersionBase:
    collection_id: CollectionId
    version_id: CollectionVersionId
    owner: str
    curator_name: str
    metadata: CollectionMetadata
    publisher_metadata: Optional[dict]  # TODO: use a dataclass
    published_at: Optional[datetime]
    created_at: datetime
    schema_version: str
    canonical_collection: CanonicalCollection

    def is_published(self) -> bool:
        """
        This collection version has been published.
        TODO: After old API code is removed consider moving closer to API layer
        """
        return self.published_at is not None and self.canonical_collection.originally_published_at is not None

    def is_unpublished_version(self) -> bool:
        """
        The collection has been published, and this is a unpublished version of the collection.
        TODO: After old API code is removed consider moving closer to API layer
        """
        return self.published_at is None and self.canonical_collection.originally_published_at is not None

    def is_initial_unpublished_version(self) -> bool:
        """
        The collection is unpublished, this version is unpublished, and no previous versions have been
        published.
        TODO: After old API code is removed consider moving closer to API layer
        """
        return self.published_at is None and self.canonical_collection.originally_published_at is None


@dataclass
class CollectionVersion(CollectionVersionBase):
    datasets: List[DatasetVersionId]


@dataclass
class CollectionVersionWithDatasets(CollectionVersionBase):
    datasets: List[DatasetVersion]


@dataclass
class CollectionVersionWithPublishedDatasets(CollectionVersionBase):
    datasets: List[PublishedDatasetVersion]
